crates/verify/src/zksync/mod.rs

```diff
@@ -0,0 +1,353 @@
+use super::{provider::VerificationProvider, EtherscanOpts, VerifyArgs, VerifyCheckArgs};
+use crate::zk_provider::{CompilerVerificationContext, ZkVerificationContext};
+use alloy_json_abi::Function;
+use alloy_primitives::hex;
+use eyre::{eyre, Result};
+use foundry_common::{abi::encode_function_args, retry::Retry};
+use foundry_compilers::zksolc::input::StandardJsonCompilerInput;
+use futures::FutureExt;
+use serde::{Deserialize, Serialize};
+use std::{fmt::Debug, thread::sleep, time::Duration};
+
+pub mod standard_json;
+
+// TODO: This is hardcoded for Sepolia verification URL, need to be updated.
+pub static ZKSYNC_VERIFICATION_URL: &str =
+    "https://explorer.sepolia.era.zksync.dev/contract_verification";
+
+#[derive(Clone, Debug, Default)]
+#[non_exhaustive]
+pub struct ZkVerificationProvider;
+
+pub trait ZksyncSourceProvider: Send + Sync + Debug {
+    fn zk_source(
+        &self,
+        context: &ZkVerificationContext,
+    ) -> Result<(StandardJsonCompilerInput, String)>;
+}
+
+#[async_trait::async_trait]
+impl VerificationProvider for ZkVerificationProvider {
+    async fn preflight_check(
+        &mut self,
+        args: VerifyArgs,
+        context: CompilerVerificationContext,
+    ) -> Result<()> {
+        let _ = self.prepare_request(&args, &context).await?;
+        Ok(())
+    }
+
+    async fn verify(
+        &mut self,
+        args: VerifyArgs,
+        context: CompilerVerificationContext,
+    ) -> Result<()> {
+        trace!("ZkVerificationProvider::verify");
+        let request = self.prepare_request(&args, &context).await?;
+
+        let client = reqwest::Client::new();
+
+        let retry: Retry = args.retry.into();
+        let verification_id: u64 = retry
+            .run_async(|| {
+                async {
+                    println!(
+                        "\nSubmitting verification for [{}] at address {}.",
+                        request.contract_name, request.contract_address
+                    );
+
+                    let verifier_url = args
+                        .verifier
+                        .verifier_url
+                        .as_deref()
+                        .unwrap_or(ZKSYNC_VERIFICATION_URL);
```

Comment:
I think we can always assume this is present, or return an error - considering that normally this is specified in the config, and overridden with the cli